app/src/main/java/com/hfad/coffeepos/data/DishDatabase.kt

```diff
@@ -2,63 +2,68 @@ package com.hfad.coffeepos.data
 import com.hfad.coffeepos.domain.usecase.DishRepository
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.catch
-import kotlinx.coroutines.flow.flow
-import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.*
 import kotlinx.coroutines.tasks.await
+import java.lang.Exception
 
 class DishDatabase(
     private val context: Context
 ) : DishRepository {
     private val auth = Firebase.auth
     private val db = Firebase.firestore
-    val batch = db.batch()
     private val dishesCollection =
         db.collection("users").document(auth.currentUser?.uid.toString())
             .collection(DISHES_DB)
 
-    override fun addDish(dish: Dish): Flow<State<String>> = flow<State<String>> {
-        val dishSh =
-            dishesCollection.whereEqualTo("name", "${dish.name}").get().await()
-        val dishes = dishSh.toObjects(Dish::class.java)
-        if (dishes.isEmpty()) {
-            dishesCollection.document(dish.name.toString()).set(dish).await()
-            emit(State.success(TRANSACTION_SUCCESS))
-        } else {
-            emit(State.failed(context.getString(R.string.error_name_dishes)))
+    override suspend fun addDish(dish: Dish): State<String> {
+        return try {
+            val dishSh =
+                dishesCollection.whereEqualTo("name", "${dish.name}").get().await()
+            val dishes = dishSh.toObjects(Ingredient::class.java)
+            if (dishes.isEmpty()) {
+                dishesCollection.document(dish.name.toString()).set(dish).await()
+                State.success(TRANSACTION_SUCCESS)
+            } else {
+                State.failed(context.getString(R.string.error_name_ingredient))
+            }
+        } catch (e: Exception) {
+            State.failed(e.message.toString())
         }
-    }.catch {
-        emit(State.failed(it.message.toString()))
-    }.flowOn(Dispatchers.IO)
+    }
 
-    override fun deleteDish(name: String) = flow<State<String>> {
-
-        dishesCollection.document(name).delete().await()
-        emit(State.success(TRANSACTION_SUCCESS))
-
-
-    }.catch {
-        emit(State.failed(it.message.toString()))
-    }.flowOn(Dispatchers.IO)
+    override suspend fun deleteDish(name: String): State<String> {
+        return try {
+            dishesCollection.document(name).delete().await()
+            State.success(TRANSACTION_SUCCESS)
+        } catch (e: Exception) {
+            State.failed(e.message.toString())
+        }
+    }
 
-    override fun observeDishes() = flow<State<List<Dish>>> {
-        val snapshot = dishesCollection.get().await()
-        val dishes = snapshot.toObjects(Dish::class.java)
-        emit(State.success(dishes))
-    }.catch {
-        emit(State.failed(it.message.toString()))
-    }.flowOn(Dispatchers.IO)
+    @ExperimentalCoroutinesApi
```

Comment:
As far as I know in the latest library version of Flow it should be NOT experimental anymore. Check, please, 1.5.0